### PR TITLE
Migrate vm-memory crate to utils

### LINKED ISSUE #3408: Move `src/vm-memory` under `src/utils`
Remove `src/vm-memory` and move the functionality under `src/utils`.

### PR DESCRIPTION
**Pull request by aaqaishtyaq**
## Changes
This PR removes the `vm-memory` crate and wraps its functionalities under `utils`. 

## Reason

Closes #3408 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by aaqaishtyaq** [comment]
Hey @JonathanWoollett-Light, Please review the PR. 

I want to run the test case suite. Can you please help approve it?

**Comment by JonathanWoollett-Light** [comment]
I think these commits can be squashed into 1 named like `` fix: Move `vm-memory` into `utils` ``

**Comment by aaqaishtyaq** [comment]
> I think these commits can be squashed into 1 named like `` fix: Move `vm-memory` into `utils` ``

Squashed the commits.

**Comment by JonathanWoollett-Light** [comment]
Nit: Git commit body could be

> Moves  `vm-memory` crate into modules in `utils` crate.

**Comment by aaqaishtyaq** [comment]
> Nit: Git commit body could be

Done

**Comment by aaqaishtyaq** [comment]
Hey @JonathanWoollett-Light, Rebased with the `main` branch. Can you please approve the CI build?

**Comment by aaqaishtyaq** [comment]
Hey @JonathanWoollett-Light, I need some help.

[This](https://github.com/aaqaishtyaq/firecracker/blob/5091ff234ea342fa02ea16841b1e5aa58ab3a36c/src/jailer/src/resource_limits.rs#L110) test case is failing, If I typecast it as `u32` then clippy is giving a warning for `musl` based builds, otherwise it is failing for both. 

If I don't typecast it then it is passing for one and failing for another as it is getting type as `i32` in one case.

```
E      --> src/jailer/src/resource_limits.rs:110:9
E       |
E   110 |         assert_eq!(u32::from(Resource::RlimitFsize), libc::RLIMIT_FSIZE as _);
E       |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-^
E       |         |                                                                  |
E       |         |                                                                  type must be known at this point
E       |         cannot infer type of the type parameter `U` declared on the function `assert_failed`
```

**Comment by aaqaishtyaq** [comment]
@JonathanWoollett-Light I have also renamed `utils` module in `src/devices/src/virtio/balloon` because `rustfmt` was adding `::utils` for the external crate and clippy was removing it, and calling the imports as unambiguous.  Using suggestion from one tool was breaking the other. Thus, I decided to rename the module from `utils` to `util`.

**Comment by aaqaishtyaq** [comment]
> Hey @JonathanWoollett-Light @roypat, I need some help.
> 
> [This](https://github.com/aaqaishtyaq/firecracker/blob/5091ff234ea342fa02ea16841b1e5aa58ab3a36c/src/jailer/src/resource_limits.rs#L110) test case is failing, If I typecast it as `u32` then clippy is giving an error for `x86_64-unknown-linux-gnu` based builds, otherwise it is failing for both.
> 


Adding error when I typecast for reference

```
E   error: casting to the same type is unnecessary (`u32` -> `u32`)
E      --> src/jailer/src/resource_limits.rs:111:55
E       |
E   111 |         assert_eq!(u32::from(Resource::RlimitNoFile), libc::RLIMIT_NOFILE as u32);
E       |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `libc::RLIMIT_NOFILE`
E       |
E       = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast

```

**Comment by JonathanWoollett-Light** [comment]
> > > Hey @JonathanWoollett-Light @roypat, I need some help.
> > > [This](https://github.com/aaqaishtyaq/firecracker/blob/5091ff234ea342fa02ea16841b1e5aa58ab3a36c/src/jailer/src/resource_limits.rs#L110) test case is failing, If I typecast it as `u32` then clippy is giving an error for `x86_64-unknown-linux-gnu` based builds, otherwise it is failing for both.
> > 
> > 
> > Adding error when I typecast for reference
> > ```
> > E   error: casting to the same type is unnecessary (`u32` -> `u32`)
> > E      --> src/jailer/src/resource_limits.rs:111:55
> > E       |
> > E   111 |         assert_eq!(u32::from(Resource::RlimitNoFile), libc::RLIMIT_NOFILE as u32);
> > E       |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `libc::RLIMIT_NOFILE`
> > E       |
> > E       = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
> > ```
> 
> Mh, I see two ways forward here, either we conditionally compile the cast, or we slap an allow for the relevant lint on it. Since there is a comment about this exact thing at the top of the file, and this is inside unit tests, I'd got with the second option.

I think it is impractical to conditional compile based on the toolchain. The ways I've seen this address in other contexts (e.g. [nix](https://github.com/nix-rust/nix)) is as you suggest @roypat to use an allow with a comment. It may also be worth opening an issue on [libc](https://github.com/rust-lang/libc) to see if this inconsistency can be addressed.

**Comment by aaqaishtyaq** [comment]
@JonathanWoollett-Light @roypat - I have added typecasting and a comment about ignoring unnecessary typecasting. 

Now, I am struggling with a failing test case. Other test cases passed when I increased the timeout to `1000ms`


```
E   failures:
E
E   ---- vstate::vcpu::x86_64::tests::test_set_tsc stdout ----
E   thread 'vstate::vcpu::x86_64::tests::test_set_tsc' panicked at 'assertion failed: vcpu.set_tsc_khz(state.tsc_khz.unwrap()).is_err()', src/vmm/src/vstate/vcpu/x86_64.rs:80
0:13
E   stack backtrace:
E      0: rust_begin_unwind
E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
E      1: core::panicking::panic_fmt
E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
E      2: core::panicking::panic
E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:114:5
E      3: core::ops::function::FnOnce::call_once
E      4: core::ops::function::FnOnce::call_once
E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/ops/function.rs:250:5
E   note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
E
E
E   failures:
E       vstate::vcpu::x86_64::tests::test_set_tsc
```

**Comment by aaqaishtyaq** [comment]
> @JonathanWoollett-Light @roypat - I have added typecasting and a comment about ignoring unnecessary typecasting.
> 
> Now, I am struggling with a failing test case (I have only updated the import paths). Other test cases passed when I increased the timeout to `1000ms`
> 
> ```
> E   failures:
> E
> E   ---- vstate::vcpu::x86_64::tests::test_set_tsc stdout ----
> E   thread 'vstate::vcpu::x86_64::tests::test_set_tsc' panicked at 'assertion failed: vcpu.set_tsc_khz(state.tsc_khz.unwrap()).is_err()', src/vmm/src/vstate/vcpu/x86_64.rs:80
> 0:13
> E   stack backtrace:
> E      0: rust_begin_unwind
> E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/std/src/panicking.rs:575:5
> E      1: core::panicking::panic_fmt
> E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:64:14
> E      2: core::panicking::panic
> E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/panicking.rs:114:5
> E      3: core::ops::function::FnOnce::call_once
> E      4: core::ops::function::FnOnce::call_once
> E                at /rustc/9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0/library/core/src/ops/function.rs:250:5
> E   note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
> E
> E
> E   failures:
> E       vstate::vcpu::x86_64::tests::test_set_tsc
> ```

@JonathanWoollett-Light @roypat - I tried running the test suite with `main` branch and the same test case failed there as well.

**Review by JonathanWoollett-Light — DISMISSED** [review]

**Comment by aaqaishtyaq** [comment]
Hey @JonathanWoollett-Light, Should I rebase with master or we can merge the branch with merge commits?

File: src/utils/Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,7 @@ thiserror = "1.0.32"
 versionize = "0.1.10"
 versionize_derive = "0.1.5"
 vmm-sys-util = "0.11.0"
+vm-memory-upstream = { package = "vm-memory",  version = "0.10.0", features = ["backend-mmap", "backend-bitmap"] }
```
Comment (roypat) [inline src/utils/Cargo.toml:16]: Mh, since there is no "vm_memory" crate in the workspace anymore, do we still need to rename the dependency?

**Comment by roypat** [comment]
> Hey @JonathanWoollett-Light, Should I rebase with main or we can merge the branch with merge commits?

GitHub will automatically drop the merge commits upon rebase-merging (you can see when you select the commit in the github view that it says "this commit merged cleanly"), so no need to worry about it :)

File: src/utils/Cargo.toml (hunk 1 above)
Comment (aaqaishtyaq) [reply src/utils/Cargo.toml]: Done. Renamed the upstream create. Please re-review.

**Review by roypat — APPROVED** [review]
Thanks for the PR! :)

**Review by JonathanWoollett-Light — APPROVED** [review]

**Comment by aaqaishtyaq** [comment]
Thanks for the reviews and for helping out during the PR cycle. @roypat @JonathanWoollett-Light

**Comment by JonathanWoollett-Light** [comment]
@aaqaishtyaq Your welcome, that's what we're here for 😅

**Pull request merged**